### PR TITLE
Possible implementation of SetThreadStackGuarantee

### PR DESCRIPTION
**Pull request by Truuum**
Maybe this implementation is not quite correct, but approximately it should look like this. If you know a better solution please let me know. Taken from the [Wine sources](https://github.com/wine-mirror/wine/blob/6e32c011a2b977246f9ab2282d30b69cff90142b/dlls/kernelbase/thread.c#L575).

### CODE REVIEW

File: src/Thunks/api-ms-win-core-processthreads.hpp [hunk 1]
```diff
@@ -158,6 +158,44 @@ namespace YY
...
+		)
+		{
+			if (auto pSetThreadStackGuarantee = try_get_SetThreadStackGuarantee())
+			{
+				return pSetThreadStackGuarantee(StackSizeInBytes);
+			}
+			else
+			{
+				ULONG prev_size = ((TEB*)NtCurrentTeb())->GuaranteedStackBytes;
+				ULONG new_size = (*StackSizeInBytes + 4095) & ~4095;
```
Comment (mingkuang-Chuyu) [inline src/Thunks/api-ms-win-core-processthreads.hpp:181]: `4095 = 4096 - 1` ? 
and in Windows 7.   it's
```cpp
(*StackSizeInBytes + GetSystemInfo().dwPageSize * 2 - 1) & ~(GetSystemInfo().dwPageSize - 1);
```

File: src/Thunks/api-ms-win-core-processthreads.hpp [hunk 2]
```diff
@@ -158,6 +158,44 @@ namespace YY
...
+				/* at least 2 pages on 64-bit */
+				if (sizeof(void*) > sizeof(int) && new_size) new_size = max(new_size, 8192);
+
+				*StackSizeInBytes = prev_size;
+				if (new_size >= (char*)((TEB*)NtCurrentTeb())->NtTib.StackBase - (char*)((TEB*)NtCurrentTeb())->DeallocationStack)
+				{
+					SetLastError(ERROR_INVALID_PARAMETER);
+					return FALSE;
+				}
+				if (new_size > prev_size) ((TEB*)NtCurrentTeb())->GuaranteedStackBytes = (new_size + 4095) & ~4095;
```
Comment (mingkuang-Chuyu) [inline src/Thunks/api-ms-win-core-processthreads.hpp:192]: Why don't we need to call the NtAllocateVirtualMemory API. Make sure the new space is always valid here?

File: src/Thunks/api-ms-win-core-processthreads.hpp (hunk 1 above)
Comment (mingkuang-Chuyu) [reply src/Thunks/api-ms-win-core-processthreads.hpp:181]: Can you tell me more useful information?

File: src/Thunks/api-ms-win-core-processthreads.hpp [hunk 3]
```diff
@@ -158,6 +158,44 @@ namespace YY
...
+			{
+				return pSetThreadStackGuarantee(StackSizeInBytes);
+			}
+			else
+			{
+				ULONG prev_size = ((TEB*)NtCurrentTeb())->GuaranteedStackBytes;
+				ULONG new_size = (*StackSizeInBytes + 4095) & ~4095;
+
+				/* at least 2 pages on 64-bit */
+				if (sizeof(void*) > sizeof(int) && new_size) new_size = max(new_size, 8192);
```
Comment (mingkuang-Chuyu) [inline src/Thunks/api-ms-win-core-processthreads.hpp:184]: 8192 = 4096 * 2 = GetSystemInfo().dwPageSize * 2

File: src/Thunks/api-ms-win-core-processthreads.hpp [hunk 4]
```diff
@@ -158,6 +158,44 @@ namespace YY
...
+
+				*StackSizeInBytes = prev_size;
+				if (new_size >= (char*)((TEB*)NtCurrentTeb())->NtTib.StackBase - (char*)((TEB*)NtCurrentTeb())->DeallocationStack)
```
Comment (mingkuang-Chuyu) [inline src/Thunks/api-ms-win-core-processthreads.hpp:187]: Sorry, I understand this code. It gives me valuable information？

File: src/Thunks/api-ms-win-core-processthreads.hpp [hunk 5]
```diff
@@ -158,6 +158,44 @@ namespace YY
...
+			_Inout_ ULONG *StackSizeInBytes
+		)
+		{
+			if (auto pSetThreadStackGuarantee = try_get_SetThreadStackGuarantee())
+			{
+				return pSetThreadStackGuarantee(StackSizeInBytes);
+			}
+			else
+			{
+				ULONG prev_size = ((TEB*)NtCurrentTeb())->GuaranteedStackBytes;
```
Comment (mingkuang-Chuyu) [inline src/Thunks/api-ms-win-core-processthreads.hpp:180]: Windows XP SP3  `TEB::GuaranteedStackBytes` (0xf78) not found, 
I only found this member `0xf78 Spare3           : Uint4B`

But you use it here, is there no problem with doing so???

**Comment by mingkuang-Chuyu** [comment]
Thanks for your PR, but I have a lot of questions about the new code and can't merge at the moment. I need to be responsible for the quality of the project, and I hope your understanding and support.

File: src/Thunks/api-ms-win-core-processthreads.hpp (hunk 1 above)
Comment (MouriNaruto) [reply src/Thunks/api-ms-win-core-processthreads.hpp:181]: @mingkuang-Chuyu I have read the implementation from WoW64's kernel32.dll in Windows Server 2003 Enterprise x64 with Service Pack 2. I found the similar logic from your said in Windows 7.

File: src/Thunks/api-ms-win-core-processthreads.hpp (hunk 3 above)
Comment (MouriNaruto) [reply src/Thunks/api-ms-win-core-processthreads.hpp:184]: @mingkuang-Chuyu The implementation from kernel32.dll in Windows Server 2003 Enterprise x86 with Service Pack 1 and the implementation from WoW64's kernel32.dll in Windows Server 2003 Enterprise x64 with Service Pack 2 **both** have the same logic. So, I don't think "at least 2 pages only on 64-bit" .

**Comment by mingkuang-Chuyu** [comment]
This PR had some issues and was not adopted, but the SetThreadStackGuarantee support code has been submitted（https://github.com/Chuyu-Team/YY-Thunks/commit/4ad4c6693d183d8e044f875bf6f8d28a09288f1c）。

**Pull request closed without merging**